juju/client/connection.py

```diff
@@ -348,12 +353,24 @@ async def _open(self, endpoint, cacert):
         else:
             url = "wss://{}/api".format(endpoint)
 
+        # We need to establish a server_hostname here for TLS sni if we are
+        # connecting through a proxy as the Juju controller certificates will
+        # not be covering the proxy
+        try:
+            sock = self.proxy.socket()
```

Comment:
I'm confused to why the exception is being thrown here. Is it because .socket() may not be on the proxy object?

Reply:
Yeah, agreed it seems very weird to catch `AttributeError` here, when you know the "shape" of the value. Is this to catch the case when `self.proxy` is None? In that case I'd use a plain `if self.proxy is not None` type of check instead of try/except.